### PR TITLE
Refactor the REST API

### PR DESCRIPTION
**Pull request by Xebax**
Refactor the REST API:

1) Modify the definition of the API so that it is closest to the commonly accepted practices. For example:
- get  the list of third parties: GET /thirdparty/list is replaced by GET /thirdparties
- create a third party: POST /thirdparty is replaced by POST /thirdparties

2) Use the Restler API Explorer for documenting the API. It allows browsing the API and also playing with it, which is a good way to learn how to use it.

### CODE REVIEW

**Comment by jfefe** [comment]
As said @Eldy the explorer sould be into an external module.

**Comment by jfefe** [comment]
To have new url for API, you just need to add doc string for it. ie : 

```
@url    GET /category/list
@url    GET /categories
```

Those 2 url will be available into API without changing anything. Please keep compatibility with changed endpoint by using doc string. 

/thirparties for POST method is a discutable choice :) With this I would tend to think that we can create several thirparties.

File: htdocs/product/class/product.class.php [hunk 1]
```diff
@@ -878,18 +878,19 @@ function update($id, $user, $notrigger=false, $action='update')
 	/**
 	 *  Delete a product from database (if not used)
 	 *
+	 * 	@param	    User	$user		Object user that ask to delete
 	 *	@param      int		$id         Product id (usage of this is deprecated, delete should be called without parameters on a fetched object)
 	 *  @param      int     $notrigger  Do not execute trigger
 	 * 	@return		int					< 0 if KO, 0 = Not possible, > 0 if OK
 	 */
-	function delete($id=0, $notrigger=0)
+	function delete($user, $id=0, $notrigger=0)
```
Comment (fappels) [inline htdocs/product/class/product.class.php]: This will break dolibarr. If you want to remove deprecated id parameter, you should replace it by the user object and ignore first parameter if int. Also better put this is separate PR.

Comment (Xebax) [inline htdocs/product/class/product.class.php]: @fappels Thank you for your feedback. Yes, you are right. I was not sure of myself when I did these changes. This is the only commit that modifies code outside of the REST API. I will revert it and fix the bug another way.

**Comment by Xebax** [comment]
@jfefe @eldy I don't understand the comment about the explorer that should be in an external module.
Dolibarr uses Restler, the code of the framework is included in Dolibarr, and the Restler API Explorer is contained in Restler. Using the Restler API Explorer only requires adding this line in the code (in htdocs/api/index.php):
`$api->r->addAPIClass('Luracast\\Restler\\Explorer');`
and then the explorer is available at https://_yourdolibarr.tld/mydolibarr_/api/index.php/explorer

**Comment by Xebax** [comment]
Here is a screenshot of the modified API browsed with the Restler API Explorer:
dolibarr - modified rest api

Please note that the "root" entry at the top is the one that contains the existing API, for backward compatibility.

**Comment by rdoursenaud** [comment]
I think that the rationale for having the explorer in a separate module is to limit Dolibarr distribution size.
This would allow packaging Dolibarr without the explorer and offer the explorer as a downloadable module on Dolistore for those who need it.
Looks reasonable enough but can be a maintenance nightmare until we have an automated way of pushing modules to the Dolistore at release time.
Anyway, it could be a sound model to distribute modules that are not strictly needed in a basic Dolibarr install but still maintain them as part of the main repository.

My 2 cents…

Great work nonetheless!

**Comment by Xebax** [comment]
@rdoursenaud I understand that distributing modules separately could be desirable. My point here is that I don't see how the Restler API Explorer - alone - could be provided in a separate module. The explorer is bundled in Restler and Restler is included in Dolibarr. It is mandatory for the REST API module.
If something must be distributed in a separate module, it would be the whole REST API, including Restler and the explorer. I do not see other option but I may have missed something.

**Comment by eldy** [comment]
A standard distribution of dolibarr is a package for USER.
RESTLER "explorer" is a tool for developper or integrator. Currently, such tools are stored into "dev" directory and they are filtered (extracted) when building final package for user.
I will make another test with you suggestion of adding a line in index.php (atfer release of 4.0). If it does not need to add libraries that is required ONLY for the dev feature, and if it does not bypass dolibarr security and work out of the box, i will reconsider my choice

**Comment by Xebax** [comment]
I don't understand the point about user VS developer. We are talking about web services here. The users of web services are developers, sure. But viewed from Dolibarr, they are still users: developing a client application that uses the web services of Dolibarr is not the same as developing inside of Dolibarr. I understand the usage of the "dev" for the internals of Dolibarr but here it is different.
When the SOAP web services are enabled, the WSDL files and the documentation are available in the configuration page.
It's the same for the REST web services: when they are available, the documentation (explorer) is available in the configuration page. My proposal is only to replace the home made explorer by the Restler API Explorer. Moreover, as explained above, it is bundled in the Restler framework, which is already included in Dolibarr. This PR does not contain any new library.

Please use this whole PR when you will make a test with the Restler API Explorer, else you will have an incomplete result.
I understand that the version 4.0 must be released before that, no problem with that.

**Comment by eldy** [comment]
I finnally solve conflict manually

**Comment by Xebax** [comment]
@eldy OK. Sorry, I got your message too late and you already had solved the conflict.
I see that the htdocs/api/admin/explorer.php file still exists after the merge but it is not needed anymore.

**Comment by eldy** [comment]
Yes. I kept it because  ode it contains help me to make test and understanding the way explorer work with a very low need of bytes. I will decide later.
About root, do you means api you find here are different than api found into other topics ? If yes, is it possible to rename it i to oldapi ?
I also have an other question: do you know how to implement a field "modified since" on api that returns full list. Goal is to add a filter on field tms >= of the request filtering date to limit answers.
